fdbclient/include/fdbclient/BlobCipher.h

```diff
@@ -181,6 +192,186 @@ struct hash<BlobCipherDetails> {
+		BinaryReader rd(flagsRef.begin(), flagsRef.size(), AssumeVersion(ProtocolVersion::withEncryptionAtRest()));
+		rd >> flags;
+		return flags;
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, encryptMode, authTokenMode, authTokenAlgo);
+	}
+};
+
+// Encryption header is stored as plaintext on a persistent storage to assist reconstruction of cipher-key(s)
+// for reads. FIPS compliance recommendation is to leverage cryptographic digest mechanism to generate
+// 'authentication token' (crypto-secure) to protect against malicious tampering and/or bit rot/flip scenarios.
+//
+// Encryption header support two modes of generation 'authentication tokens':
+// 1) SingleAuthTokenMode: the scheme generates single crypto-secrure auth token to protect {cipherText +
+// header} payload. Scheme is geared towards optimizing cost due to crypto-secure auth-token generation,
+// however, on decryption client needs to be read 'header' + 'encrypted-buffer' to validate the 'auth-token'.
+// The scheme is ideal for usecases where payload represented by the encryptionHeader is not large and it is
+// desirable to minimize CPU/latency penalty due to crypto-secure ops, such as: CommitProxies encrypted inline
+// transactions, StorageServer encrypting pages etc.
+// SOMEDAY: Another potential scheme could be 'MultiAuthTokenMode': Scheme generates separate authTokens
+// for 'encrypted buffer' & 'encryption-header'. The scheme is ideal where payload represented by
+// encryptionHeader is large enough such that it is desirable to optimize cost of upfront reading full
+// 'encrypted buffer', compared to reading only encryptionHeader and ensuring its sanity; for instance:
+// backup-files.
+
+template <uint32_t AuthTokenSize>
+struct AesCtrWithAuthV1 {
+	// Serializable fields
+
+	// Text cipher encryption information
+	BlobCipherDetails cipherTextDetails;
+	// Header cipher encryption information
+	BlobCipherDetails cipherHeaderDetails;
+	// Initialization vector
+	uint8_t iv[AES_256_IV_LENGTH];
+	// Authentication token
+	uint8_t authToken[AuthTokenSize];
+
+	AesCtrWithAuthV1() {}
+	AesCtrWithAuthV1(const BlobCipherDetails& textDetails,
+	                 const BlobCipherDetails& headerDetails,
+	                 const uint8_t* ivBuf,
+	                 const int ivLen,
+	                 Arena& arena)
+	  : cipherTextDetails(textDetails), cipherHeaderDetails(headerDetails) {
+		ASSERT_EQ(ivLen, AES_256_IV_LENGTH);
+		memcpy(&iv[0], ivBuf, ivLen);
+		memset(&authToken[0], 0, AuthTokenSize);
+	}
+
+	bool operator==(const AesCtrWithAuthV1<AuthTokenSize>& o) const {
+		return cipherHeaderDetails == o.cipherHeaderDetails && cipherTextDetails == o.cipherTextDetails &&
+		       memcmp(&iv[0], &o.iv[0], AES_256_IV_LENGTH) == 0 &&
+		       memcmp(&authToken[0], &o.authToken[0], AuthTokenSize) == 0;
+	}
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		ar.serializeBytes(this, sizeof(AesCtrWithAuthV1<AuthTokenSize>));
```

Comment:
nit: can we do this? the difference is small but it will allow evolving together with `BlobCipherDetails`'s serializer. Same for other similar structs.
```
template <class Ar>
void serialize(Ar& ar) {
    serializer(ar, cipherTextDetails, cipherHeaderDetails);
    ar.serializeBytes(iv, AES_256_IV_LENGTH);
    ar.serializeBytes(authToken, AuthTokenSize);
}
```

Reply:
> the difference is small
First, sure we can do it; however, am not sure if there is any difference. For future evolution, `V1` tag shouldn't be evolved for any major changes, for instance one advantage of having `algoHeaderVersion` is, even for same EncodingScheme the corresponding AlgorithmHeader can evolve and at that point we should define `V2` for more clarity (should be rare event, but if it happens).

However, the suggested code looks more readable and I like the idea of adopting to this format (more explicit).
Thanks